### PR TITLE
feat(openai-enhancements): improved user experience when interacting with openai

### PR DESCRIPTION
**Pull request by boltdynamics**
### Replies to questions on the slack post's thread
Screenshot 2023-07-26 at 8 44 49 pm

### If you ask a question from inside a thread, replies in the thread
image

### If you edit a post, also replies in the thread
image

### The same issue where Slack expects a response within 3 seconds of the request. We decoupled the bot and openai functions and put a cloud task queue in between to allow the first function to respond to Slack immediately. But even the first function can take longer than 3 seconds at times where you see the bot responding multiple times as seen in the screenshot below.
image

@geshan @sonam-serchan @Aaasisss @pgaijin66 Let me know if you have any thoughts for above. Any issues if we move the application to AWS? AWS API Gateway can solve this - responds to Slack immediately with a 200 OK and then asynchronously invokes the lambda function. Our Golang version of the bot already does it here https://github.com/xplorer-io/xplorers-bot-go

### CODE REVIEW

**Comment by geshan** [comment]
@boltdynamics GCP also has an API gateway: https://cloud.google.com/api-gateway but if you think AWS makes more sense switch it out :)

File: package.json [hunk 1]
```diff
@@ -3,7 +3,8 @@
   "scripts": {
     "start": "functions-framework --target=xplorersbot",
     "test": "task clean && jest --verbose --coverage tests",
-    "start-local-server": "task clean && tsc && cd out && npx functions-framework --target=xplorersbot --signature-type=http"
+    "start-xplorersbot-server": "task clean && tsc && cd out && npx functions-framework --target=xplorersbot --signature-type=http",
```
Comment (geshan) [inline package.json:6]: it is a bit strange that package-lock.json changed without changing any dependencies. not a blocker though.

**Comment by boltdynamics** [comment]
> @boltdynamics GCP also has an API gateway: https://cloud.google.com/api-gateway but if you think AWS makes more sense switch it out :)

From my initial research, I don't think you can asynchronously invoke a Google cloud function using GCP API Gateway like we can with AWS API Gateway. Correct me if I am wrong, and if there is a way to make this work in GCP, I'd still prefer to keep things the way they are now in GCP.

Let me know what you think 💬

File: src/helpers/slack.ts [hunk 2]
```diff
@@ -141,15 +141,21 @@ export async function addReactionToSlackPost(
...
+            text: text,
+            channel: slackChannel,
+            thread_ts: thread_ts,
+        });
+    } else {
+        return await slackWebClient.chat.postMessage({
+            text: text,
+            channel: slackChannel,
+        });
+    }
```
Comment (geshan) [inline src/helpers/slack.ts:158]:
```suggestion
    let message = {
      text,
      channel: slackChannel
    };
    
    if (threadTs) { //for consistency make it camelCase like slackChannel :)
      message.thread_ts = threadTs;
    }
    
        return await slackWebClient.chat.postMessage(message);
    }
```

Else can be avoided 90% of the time :).

File: src/index.ts [hunk 3]
```diff
@@ -41,26 +41,39 @@ export const xplorersbot: HttpFunction = async (req, res) => {
...
-            const messageStartsWithHeyOpenAI = slackEvent?.text
-                .toLowerCase()
-                .startsWith("hey openai");
 
-            if (isChannelOpenAI && messageStartsWithHeyOpenAI) {
-                await createHttpTask(req.body);
+            // text could be in slackEvent.text or slackEvent.message.text
+            const message = slackEvent?.text ?? slackEvent?.message?.text;
+
+            if (isChannelOpenAI) {
```
Comment (geshan) [inline src/index.ts:58]: can be done in 1 level if `if(isChannelOpenAI && messageStartsWithHeyOpenAI)`... 

Up to you to decide :).

**Comment by geshan** [comment]
> > @boltdynamics GCP also has an API gateway: https://cloud.google.com/api-gateway but if you think AWS makes more sense switch it out :)
> 
> From my initial research, I don't think you can asynchronously invoke a Google cloud function using GCP API Gateway like we can with AWS API Gateway. Correct me if I am wrong, and if there is a way to make this work in GCP, I'd still prefer to keep things the way they are now in GCP.
> 
> Let me know what you think 💬

Not sure of that might need to experiment a bit. If it doesn't work move to AWS.

**Review by geshan — APPROVED** [review]

File: src/helpers/slack.ts (hunk 2 above)
Comment (boltdynamics) [reply src/helpers/slack.ts]: Thank you @geshan , I did remove the else block 😍 

I also added a type to message var.

File: src/index.ts (hunk 3 above)
Comment (boltdynamics) [reply src/index.ts]: Indeed, changed it to be 1 level 👍

**Comment by boltdynamics** [comment]
> > > @boltdynamics GCP also has an API gateway: https://cloud.google.com/api-gateway but if you think AWS makes more sense switch it out :)
> > 
> > 
> > From my initial research, I don't think you can asynchronously invoke a Google cloud function using GCP API Gateway like we can with AWS API Gateway. Correct me if I am wrong, and if there is a way to make this work in GCP, I'd still prefer to keep things the way they are now in GCP.
> > Let me know what you think 💬
> 
> Not sure of that might need to experiment a bit. If it doesn't work move to AWS.

I will do some more research, will only move to AWS if we are unable to find an optimal solution in GCP

**Review by geshan — APPROVED** [review]

File: src/index.ts [hunk 4]
```diff
@@ -59,7 +59,7 @@ export const xplorersbot: HttpFunction = async (req, res) => {
                 .toLowerCase()
                 .startsWith("hey openai");
 
-            if (slackEvent?.type === "message") {
+            if (!isChannelOpenAI && slackEvent?.type === "message") {
```
Comment (boltdynamics) [inline src/index.ts:62]: @geshan Sorry I had to this to avoid processing the event in openai channel and to avoid welcoming users in that channel

Comment (geshan) [reply src/index.ts:62]: All good

**Pull request merged**